### PR TITLE
PostgreSQL database driver

### PR DESCRIPTION
**Pull request by BenLubar**
because the world needs more confusing SQL schemas

### CODE REVIEW

File: install/databases.js [hunk 1]
```diff
@@ -69,11 +76,19 @@ function saveDatabaseConfig(config, databaseConfig, callback) {
...
+			username: databaseConfig['postgres:username'],
+			password: databaseConfig['postgres:password'],
+			database: databaseConfig['postgres:database'],
+		};
 	} else {
 		return callback(new Error('unknown database : ' + config.database));
 	}
 
-	var allQuestions = questions.redis.concat(questions.mongo);
+	var allQuestions = questions.redis.concat(questions.mongo).concat(questions.postgres);
```
Comment (pitaj) [inline install/databases.js]: `.concat(questions.mongo, questions.postgres)`

Comment (BenLubar) [inline install/databases.js]: `git grep ')\.concat(' src` returns quite a few places where the `.concat().concat()` style is used.

Comment (pitaj) [inline install/databases.js]: You're right... but still tho

**Comment by BenLubar** [comment]
@julianlam @barisusakli @psychobunny When https://github.com/barisusakli/nodebb-plugin-dbsearch/pull/35 gets merged, feel free to push the plugin version bump to my branch so Travis can test it. It works locally, but just to be safe...

**Comment by barisusakli** [comment]
@BenLubar dbsearch 2.0.7, hopefully the tests were useful. Writing the mongodb driver with no tests was no fun 😈

**Comment by BenLubar** [comment]
Hmm, all tests pass, but I'm getting this when loading the front page of the forum:

```
9/8 07:01:46 [10017] - error: [4568/10017]  TypeError: Cannot create property 'start' on number '32'
    at async.waterfall.totalPinnedCount (/usr/src/app/src/categories/topics.js:46:25)
    at nextTask (/usr/src/app/node_modules/async/dist/async.js:5273:14)
    at Object.waterfall (/usr/src/app/node_modules/async/dist/async.js:5283:5)
    at Categories.getTopicIds (/usr/src/app/src/categories/topics.js:43:9)
    at /usr/src/app/node_modules/async/dist/async.js:1268:19
    at /usr/src/app/node_modules/async/dist/async.js:3830:24
    at replenish (/usr/src/app/node_modules/async/dist/async.js:946:17)
    at /usr/src/app/node_modules/async/dist/async.js:950:9
    at eachOfLimit (/usr/src/app/node_modules/async/dist/async.js:975:24)
    at /usr/src/app/node_modules/async/dist/async.js:980:16
    at eachOf (/usr/src/app/node_modules/async/dist/async.js:1051:5)
    at _parallel (/usr/src/app/node_modules/async/dist/async.js:3829:5)
    at Object.parallelLimit [as parallel] (/usr/src/app/node_modules/async/dist/async.js:3912:5)
    at Object.header (/usr/src/app/node_modules/nodebb-plugin-tdwtf-customizations/index.js:308:9)
    at /usr/src/app/src/plugins/hooks.js:123:12
    at /usr/src/app/node_modules/async/dist/async.js:2488:9
    at replenish (/usr/src/app/node_modules/async/dist/async.js:946:17)
    at /usr/src/app/node_modules/async/dist/async.js:950:9
    at eachOfLimit (/usr/src/app/node_modules/async/dist/async.js:975:24)
    at /usr/src/app/node_modules/async/dist/async.js:980:16
    at Object.reduce (/usr/src/app/node_modules/async/dist/async.js:2487:5)
    at fireFilterHook (/usr/src/app/src/plugins/hooks.js:115:9)
    at Object.Plugins.fireHook (/usr/src/app/src/plugins/hooks.js:96:4)
    at /usr/src/app/src/middleware/header.js:162:13
    at nextTask (/usr/src/app/node_modules/async/dist/async.js:5273:14)
    at next (/usr/src/app/node_modules/async/dist/async.js:5280:9)
    at /usr/src/app/node_modules/async/dist/async.js:906:16
    at /usr/src/app/node_modules/async/dist/async.js:3838:9
    at /usr/src/app/node_modules/async/dist/async.js:421:16
    at iterateeCallback (/usr/src/app/node_modules/async/dist/async.js:928:24)
    at /usr/src/app/node_modules/async/dist/async.js:906:16
    at /usr/src/app/node_modules/async/dist/async.js:3835:13
    at /usr/src/app/src/user.js:203:3
    at /usr/src/app/node_modules/async/dist/async.js:421:16
    at next (/usr/src/app/node_modules/async/dist/async.js:5278:29)
    at /usr/src/app/node_modules/async/dist/async.js:906:16
    at /usr/src/app/src/user.js:336:4
    at nextTask (/usr/src/app/node_modules/async/dist/async.js:5273:14)
    at next (/usr/src/app/node_modules/async/dist/async.js:5280:9)
    at /usr/src/app/node_modules/async/dist/async.js:906:16
    at /usr/src/app/node_modules/async/dist/async.js:421:16
    at next (/usr/src/app/node_modules/async/dist/async.js:5278:29)
    at /usr/src/app/node_modules/async/dist/async.js:906:16
    at /usr/src/app/node_modules/async/dist/async.js:421:16
    at next (/usr/src/app/node_modules/async/dist/async.js:5278:29)
    at /usr/src/app/node_modules/async/dist/async.js:906:16
    at /usr/src/app/src/privileges/users.js:120:5
    at nextTask (/usr/src/app/node_modules/async/dist/async.js:5273:14)
    at next (/usr/src/app/node_modules/async/dist/async.js:5280:9)
    at /usr/src/app/node_modules/async/dist/async.js:906:16
    at /usr/src/app/src/plugins/hooks.js:129:4
    at /usr/src/app/node_modules/async/dist/async.js:421:16
    at /usr/src/app/node_modules/async/dist/async.js:2493:9
    at /usr/src/app/node_modules/async/dist/async.js:421:16
    at replenish (/usr/src/app/node_modules/async/dist/async.js:941:25)
    at iterateeCallback (/usr/src/app/node_modules/async/dist/async.js:931:17)
    at /usr/src/app/node_modules/async/dist/async.js:906:16
    at /usr/src/app/node_modules/async/dist/async.js:2490:13
    at /usr/src/app/node_modules/nodebb-plugin-deny-moderator-access/index.js:38:3
    at /usr/src/app/node_modules/nodebb-plugin-deny-moderator-access/index.js:55:3
    at /usr/src/app/node_modules/async/dist/async.js:421:16
    at next (/usr/src/app/node_modules/async/dist/async.js:5278:29)
    at /usr/src/app/node_modules/async/dist/async.js:906:16
    at /usr/src/app/src/privileges/helpers.js:72:4
    at nextTask (/usr/src/app/node_modules/async/dist/async.js:5273:14)
    at next (/usr/src/app/node_modules/async/dist/async.js:5280:9)
    at /usr/src/app/node_modules/async/dist/async.js:906:16
    at /usr/src/app/node_modules/async/dist/async.js:3838:9
    at /usr/src/app/node_modules/async/dist/async.js:421:16
    at iterateeCallback (/usr/src/app/node_modules/async/dist/async.js:928:24)
    at /usr/src/app/node_modules/async/dist/async.js:906:16
    at /usr/src/app/node_modules/async/dist/async.js:3835:13
    at /usr/src/app/node_modules/async/dist/async.js:421:16
    at next (/usr/src/app/node_modules/async/dist/async.js:5278:29)
    at /usr/src/app/node_modules/async/dist/async.js:906:16
    at /usr/src/app/src/groups/membership.js:486:5
    at nextTask (/usr/src/app/node_modules/async/dist/async.js:5273:14)
    at next (/usr/src/app/node_modules/async/dist/async.js:5280:9)
    at /usr/src/app/node_modules/async/dist/async.js:906:16
    at /usr/src/app/node_modules/async/dist/async.js:421:16
    at Immediate.next (/usr/src/app/node_modules/async/dist/async.js:5278:29)
    at Immediate.<anonymous> (/usr/src/app/node_modules/async/dist/async.js:906:16)
    at runCallback (timers.js:676:20)
    at tryOnImmediate (timers.js:645:5)
    at processImmediate [as _immediateCallback] (timers.js:617:5)
```

**Comment by julianlam** [comment]
... oh dear, it seems our robot hasn't signed the CLA 😆

**Comment by BenLubar** [comment]
@julianlam oops, I rebased onto master instead of develop

**Comment by BenLubar** [comment]
Anyway, I just wanted to run the test cases with up-to-date NodeBB code before I tried to add transactions to the database package.

**Comment by BenLubar** [comment]
@barisusakli Travis died between my last two commits: https://www.traviscistatus.com/incidents/yxqx6brg5nl0

The tests all pass, but there is a sporadic error that causes `acp.min.js` to not be built, and I'm not sure what's causing that, but it's probably unrelated to this PR.

File: src/database/postgres/hash.js [hunk 2]
```diff
@@ -46,7 +50,11 @@ module.exports = function (db, module) {
 				async.apply(helpers.ensureLegacyObjectType, tx.client, key, 'hash'),
 				async.apply(query, {
 					name: 'setObjectField',
-					text: 'INSERT INTO "legacy_hash" ("_key", "data") VALUES ($1::TEXT, jsonb_build_object($2::TEXT, $3::TEXT::JSONB)) ON CONFLICT ("_key") DO UPDATE SET "data" = jsonb_set("legacy_hash"."data", ARRAY[$2::TEXT], $3::TEXT::JSONB)',
+					text: `
+INSERT INTO "legacy_hash" ("_key", "data")
```
Comment (pitaj) [inline src/database/postgres/hash.js]: Any reason these aren't indented evenly? I didn't think SQL had significant whitespace.

Comment (BenLubar) [reply src/database/postgres/hash.js]: I'm using the same format I use for SQL at work, where the first keyword per line is right-aligned with other things at the same nesting level.

**Comment by BenLubar** [comment]
[rebased onto master]

**Comment by hmqgg** [comment]
I merged it in production, and it works well. Performance loss is not significant.
Although it needs to add monitoring and more features. I'm working on it.
Deployment: Docker.

**Comment by venmade** [comment]
In what state is the introduction of PostgreSQL?

**Comment by julianlam** [comment]
:+1: Okay to merge. There's no particular blocker on our end, just when you decide it is ready. Unfortunately _supporting_ these kinds of installs are a separate matter entirely, but we'll need to merge it in first and then see what kind of issues we'll encounter over time :smile:

**Comment by BenLubar** [comment]
@julianlam if you want to wait until tomorrow, WTDWTF will be down from 2018-07-08T05:00:00Z to about 2018-07-08T06:00:00Z to convert to PostgreSQL.

**Comment by BenLubar** [comment]
It's now live in production and there have been no major problems:

https://what.thedailywtf.com/post/1374386

**Comment by BenLubar** [comment]
rebased

**Pull request merged**